### PR TITLE
8261625: Add `Elements.isAutomaticModule(ModuleElement)`

### PR DESCRIPTION
**Pull request by jddarcy**
Addition of a small convenience method, with a default implementation, to Elements.

Most of the change is testing infrastructure. The negative cases, such as named modules, are tested in the style of other annotation processing tests for Elements utilities.

An automatic module comes from a jar file, so an API reading from a jar file is needed. To avoid piping in the awkward plumbing needed to do that directly in the annotation processing tests, I updated an existing automatic modules test which uses the tool box facility. My additions to the tool box might not be the done in the most idiomatic style for that API; happy to have guidance on how to proceed better there.

Assuming this general approach is agreed to, I'll update the imports on AutomaticModules.java to condense the text of the annotation processor source.

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8261625](https://bugs.openjdk.java.net/browse/JDK-8261625): Add `Elements.isAutomaticModule(ModuleElement)`

### Reviewers
 * [Vicente Romero](https://openjdk.java.net/census#vromero) (@vicente-romero-oracle - **Reviewer**) ⚠️ Review applies to 5dab5c2fbcdee9e85f49666ae91c3ae3406e1912
 * [Jan Lahoda](https://openjdk.java.net/census#jlahoda) (@lahodaj - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk pull/3382/head:pull/3382` \
`$ git checkout pull/3382`

Update a local copy of the PR: \
`$ git checkout pull/3382` \
`$ git pull https://git.openjdk.java.net/jdk pull/3382/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 3382`

View PR using the GUI difftool: \
`$ git pr show -t 3382`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk/pull/3382.diff">https://git.openjdk.java.net/jdk/pull/3382.diff</a>

</details>

### CODE REVIEW

**Comment by jddarcy** [comment]
PS And the CSR should be reviewed as well:

https://bugs.openjdk.java.net/browse/JDK-8264865

Thanks.

File: test/langtools/tools/javac/modules/AutomaticModules.java [hunk 1]
```diff
@@ -665,7 +665,31 @@ public void testAutomaticModuleNameCorrect(Path base) throws Exception {
...
+                                 javax.lang.model.util.Elements elts = processingEnv.getElementUtils();
+                                 if (!elts.isAutomaticModule(elts.getModuleOf(elts.getTypeElement("api.Api")))) {
+                                     throw new RuntimeException("module of class Api is not automatic");
+                                 }
+                             }
+                             return true;
+                         }
+                     })
                 .writeAll()
                 .getOutputLines(Task.OutputKind.DIRECT);
```
Comment (vicente-romero-oracle) [inline test/langtools/tools/javac/modules/AutomaticModules.java:694]: I know this code predates your patch but now that you are here :) these two lines are unnecessary as the output is not kept. I mean you can remove:
           .writeAll()
           .getOutputLines(Task.OutputKind.DIRECT)

File: test/langtools/tools/javac/processing/model/util/elements/TestIsAutomaticMod.java [hunk 2]
```diff
@@ -0,0 +1,125 @@
+/*
+ * Copyright (c) 2006, 2021, Oracle and/or its affiliates. All rights reserved.
```
Comment (vicente-romero-oracle) [inline test/langtools/tools/javac/processing/model/util/elements/TestIsAutomaticMod.java:2]: 2006 can be removed

File: test/langtools/tools/lib/toolbox/JavacTask.java [hunk 3]
```diff
@@ -331,6 +333,29 @@ public String name() {
         return checkExit(new Task.Result(toolBox, this, rc, outputMap));
     }
 
+    /**
+     * Sets the expected outcome of the task and calls {@code run(procs)}.
+     * @param expect the expected outcome
+     * @return the result of calling {@code run}
+     */
+    public Result run(Expect expect, Processor... procs) {
```
Comment (vicente-romero-oracle) [inline test/langtools/tools/lib/toolbox/JavacTask.java:341]: I would prefer to add a new method to this class, something in the lines of:
```
public JavacTask processors(Processor... procs) {
    this.procs = List.of(procs);
    return this;
}
```

which I think is more in the spirit of the existing API, then you won't need to add the two new versions of the `run` method

**Review by vicente-romero-oracle — CHANGES_REQUESTED** [review]

File: test/langtools/tools/javac/processing/model/util/elements/TestIsAutomaticMod.java (hunk 2 above)
Comment (jddarcy) [reply test/langtools/tools/javac/processing/model/util/elements/TestIsAutomaticMod.java:2]: I copied a file from 2006 and then edited the result, hence I started with the 2006 year.

File: test/langtools/tools/lib/toolbox/JavacTask.java (hunk 3 above)
Comment (jddarcy) [reply test/langtools/tools/lib/toolbox/JavacTask.java]: Added as suggested; thanks.

File: test/langtools/tools/javac/processing/model/util/elements/TestIsAutomaticMod.java [hunk 4]
```diff
@@ -0,0 +1,125 @@
...
+                           RoundEnvironment roundEnv) {
+        if (!roundEnv.processingOver()) {
+            // Named module java.base
+            checkMod(eltUtils.getModuleElement("java.base"), false);
+
+            // Unnamed module for TestIsAutomaticMod
+            for (Element e : roundEnv.getRootElements() ) {
+                ModuleElement enclosing = elements.getModuleOf(e);
+                checkMod(enclosing, false);
+                System.out.println(enclosing.toString());
```
Comment (vicente-romero-oracle) [inline test/langtools/tools/javac/processing/model/util/elements/TestIsAutomaticMod.java:53]: can this line `System.out...` be removed? probably a debug leftover?

**Review by vicente-romero-oracle — APPROVED** [review]
lgtm

File: test/langtools/tools/javac/processing/model/util/elements/TestIsAutomaticMod.java (hunk 4 above)
Comment (jddarcy) [reply test/langtools/tools/javac/processing/model/util/elements/TestIsAutomaticMod.java:53]: Sure; I'll remove it before pushing.

**Review by lahodaj — APPROVED** [review]
LGTM

**Pull request closed without merging**